### ISSUE TITLE
Add ExceptionSorter support

### ISSUE THREAD

**Issue by yaojuncn** [brettwooldridge/HikariCP#387]
It looks to me that when ConnectionProxy checks for SQLExceptions from database, 
it only checks limited static sql errors to decide whether the error is fatal enough to close the connection; 

when different db & driver vender may implement it quite differently; and some times the drivers have bugs of providing the correct code, especially under some different OS or so on;

good connection pool implementations usually provide the support for allowing customized exceptionsorter so that it can allow users to customize the SQL errors that will be marked as 'fatal' and connections should be closed;

I'd suggest HikariCP also add this support;
Correct me if you already did that;

**Comment by brettwooldridge** [comment]
Vibur DBCP recently [implemented it](http://www.vibur.org/#critical-sql-states). Maybe those codes look familiar?  I wonder where they got them. :grinning:

In general, we try very hard not to add additional configuration options.  Our approach has been, and likely will continue to be, to address specific issues as they are raised by the community.

Rather than having a new configuration option, I would rather increase the list of SQLState codes whenever one is reported as missing.

By specification SQLState codes with the "08" prefix represent connection failures, and perhaps surprisingly most vendors generally follow the spec.  Vendors are free to create their own variety of states *within the "08" prefix*, and they do.  We have already read the documentation of most of the major databases.  There are a few "outliers", but very few.

In summary, before adding a new configuration property that 99.7% of users will never care about, I'd rather sit back and wait for users to report issues *(but they haven't in the past 2 years)*.

**Comment by nitincchauhan** [comment]
Something to note about:
HikariCP marks 'that' connection for closing for such exceptions.
but interestingly Vibur, as doc says will close ALL connection from pool on such exception!

**Comment by brettwooldridge** [comment]
Yes, this is noted in our [Pool Analysis](https://github.com/brettwooldridge/HikariCP/wiki/Pool-Analysis#exception-detection).

**Comment by yaojuncn** [comment]
1. if you know the product has a potential issue and do not fix it until someone reported a bug; this is putting your users in danger; if it's some web community site it's fine; but some users may be using it in serious money trading applications, any thing that's known as a potential issue, is a serious issue; 
these users would prefer to prevent issues rather than let it happen and then fix; 

elegant is nice but nothing is more important than correctness. please consider;
2. 
JBoss connection pool 4.0 (which is about 10 years ago), has the feature of exceptionsorter; 
my own connection pool implementation, YDataSource, which is done two years ago in a weekend also support this; 
https://github.com/yaojuncn/YDataSource/blob/master/src/YDataSource-Core/src/main/java/com/yaojun/jdbc/ExceptionSorter.java
And I support dynamic setting the error code in runtime; 
Druid, which is a pretty cool connection pool as well, also support this feature;
so it's nothing new, it's actually some basic feature;

3. close ALL connections is not the right behavior in terms of connection pool management; one connection has broken does not mean all connections needs to be closed, although it's likely to be the case, but not always. closing all connections will be a huge damage to applications in some cases;

**Comment by brettwooldridge** [comment]
If an device has a disconnection error, even if not detected from the SQLException, the connection will be ejected from the pool when tested.  In HikariCP, disconnection detection via SQLState is merely an optimization, not a correctness issue.

The reason Vibur and other pools may need a configurable list of codes is that they do not test connections on borrow by default.  Therefore the risk is high for them if they miss a detection.

You could delete detection entirely from HikariCP and correctness would not be affected.
